spec/java/security/SecureRandom.main.lsl

```diff
@@ -0,0 +1,885 @@
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    constructor *.SecureRandom (@target self: SecureRandom, seed: array<byte>)
+    {
+        _getDefaultPRNG(false, null);
+    }
+
+
+    // static methods
+
+    @throws(["java.security.NoSuchAlgorithmException"])
+    @static fun *.getInstance (_algorithm: String): SecureRandom
+    {
+        val providersList: array<Provider> = action CALL_METHOD(null as Security, "getProviders", []);
+        val providersListLength: int = action ARRAY_SIZE(providersList);
+        var resultProvider: Provider = null;
+        var resultAlgorithm: String = null;
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, providersListLength, +1,
+            findAlgorithmByAllProviders_loop(i, providersList, _algorithm, resultProvider, resultAlgorithm)
+        );
+
+        if (resultAlgorithm == null)
+            _throwNSAE();
+
+        val isDefaultProvider: boolean = _isDefaultProvider(resultProvider);
+
+        result = new SecureRandomAutomaton(state = Initialized,
+            provider = resultProvider,
+            algorithm = resultAlgorithm,
+            defaultProvider = isDefaultProvider,
+        );
+    }
+
+
+    @Phantom proc findAlgorithmByAllProviders_loop (i: int, providersList: array<Provider>, _algorithm: String, resultProvider: Provider, resultAlgorithm: String): void
+    {
+        val curProvider: Provider = providersList[i];
+        val services: Set = action CALL_METHOD(curProvider, "getServices", []);
+        val iter: Iterator = action CALL_METHOD(services, "iterator", []);
+        resultProvider = curProvider;
+
+        action LOOP_WHILE(
+            action CALL_METHOD(iter, "hasNext", []),
+            findAlgorithm_loop(iter, curProvider, _algorithm, resultAlgorithm)
+        );
+    }
+
+
+    @Phantom proc findAlgorithm_loop (iter: Iterator, curProvider: Provider, _algorithm: String, resultAlgorithm: String): void
+    {
+        val curService: Provider_Service = action CALL_METHOD(iter, "next", []) as Provider_Service;
+        val curServiceType: String = action CALL_METHOD(curService, "getType", []);
+        val curServiceAlgorithm: String = action CALL_METHOD(curService, "getAlgorithm", []);
+
+        if (action OBJECT_EQUALS(curServiceType, "SecureRandom") && action OBJECT_EQUALS(curServiceAlgorithm, _algorithm))
+        {
+            resultAlgorithm = curServiceAlgorithm;
+            action LOOP_BREAK();
+        }
+    }
+
+
+    @throws(["java.security.NoSuchAlgorithmException"])
+    @static fun *.getInstance (_algorithm: String, provider: Provider): SecureRandom
+    {
+        if (provider == null)
+            _throwIAE();
+
+        val providersListLength: int = 1;
+        val providersList: array<Provider> = action ARRAY_NEW("java.security.Provider", providersListLength);
+        providersList[0] = provider;
```

Comment:
@dog-m 
This lines were deleted, because were used:

```
action SYMBOLIC("java.security.Provider");
action SYMBOLIC("java.lang.String");
```

temporary decision;